File: RxUI/src/main/java/com/vondear/rxui/view/scaleimage/decoder/CompatDecoderFactory.java
package com.vondear.rxui.view.scaleimage.decoder;


import androidx.annotation.NonNull;

/**
 * @param <T> The base type of the decoder this factory will produce.
 *            <p>
 *            Compatibility factory to instantiate decoders with empty public constructors.
 * @author vondear
 */
public class CompatDecoderFactory<T> implements DecoderFactory<T> {
    private Class<? extends T> clazz;

    public CompatDecoderFactory(@NonNull Class<? extends T> clazz) {
        this.clazz = clazz;
    }

    @Override
    public T make() throws IllegalAccessException, InstantiationException {
        return clazz.newInstance();
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/scaleimage/decoder/CompatDecoderFactory.kt
package com.tamsiree.rxui.view.scaleimage.decoder

/**
 * @param <T> The base type of the decoder this factory will produce.
 *
 *
 * Compatibility factory to instantiate decoders with empty public constructors.
 * @author tamsiree
</T> */
class CompatDecoderFactory<T>(private val clazz: Class<out T>) : DecoderFactory<T> {
    @Throws(IllegalAccessException::class, InstantiationException::class)
    override fun make(): T {
        return clazz.newInstance()
    }

}